// react-multiselect-checkboxes/lib/CheckboxGroup.jsx
import React from 'react';
import PropTypes from 'prop-types';
import { css } from 'emotion';
import CheckboxWithIndeterminate from './CheckboxWithIndeterminate';

function CheckboxGroup(props) {
  const { children, className, cx, getStyles, Heading, setValue, data, label, getValue } = props;
  const viableOptions = data.options;
  const numCheckedOptions = getValue().filter((v) => viableOptions.includes(v)).length;
  const checked = numCheckedOptions === data.options.length;
  const indeterminate = numCheckedOptions > 0 && !checked;
  const checkboxProps = { indeterminate, checked };
  const selectAllOptions = () => {
    const newValue = [...getValue().filter((v) => !data.options.includes(v)), ...data.options];
    setValue(newValue);
  };
  const clearOptions = () => {
    const newValue = getValue().filter((v) => !data.options.includes(v));
    setValue(newValue);
  };
  return (
    <div className={cx(css(getStyles('group', props)), { group: true }, className)}>
      <Heading
        checkboxProps={checkboxProps}
        getStyles={getStyles}
        cx={cx}
        onClick={() => {
          if (indeterminate || checked) {
            clearOptions();
          } else {
            selectAllOptions();
          }
        }}
      >
        {label}
      </Heading>
      <div>{children}</div>
    </div>
  );
}

CheckboxGroup.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string.isRequired,
  cx: PropTypes.func.isRequired,
  getStyles: PropTypes.func.isRequired,
  Heading: PropTypes.func.isRequired,
  getValue: PropTypes.func.isRequired,
  setValue: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  data: PropTypes.shape({
    options: PropTypes.arrayOf(
      PropTypes.shape({
        label: PropTypes.node,
        value: PropTypes.any,
      }),
    ),
  }).isRequired,
};
CheckboxGroup.defaultProps = {
  children: null,
};

export function CheckboxGroupHeading(props) {
  const { className, cx, getStyles, children, checkboxProps, ...cleanProps } = props;
  return (
    <div
      className={cx(
        css(getStyles('groupHeading', { ...props, ...checkboxProps })),
        { 'group-heading': true },
        className,
      )}
      {...cleanProps}
    >
      <CheckboxWithIndeterminate readOnly type="checkbox" {...checkboxProps} />
      {children}
    </div>
  );
}

CheckboxGroupHeading.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string.isRequired,
  cx: PropTypes.func.isRequired,
  getStyles: PropTypes.func.isRequired,
  Heading: PropTypes.func.isRequired,
  getValue: PropTypes.func.isRequired,
  setValue: PropTypes.func.isRequired,
  label: PropTypes.string.isRequired,
  checkboxProps: PropTypes.shape({
    checked: PropTypes.bool,
    indeterminate: PropTypes.bool,
  }).isRequired,,
};
CheckboxGroupHeading.defaultProps = {
  children: null,
};

export default CheckboxGroup;
